Add an HTTP/2 304 "Not Modified" AuTest.

bneradt opened the pull request:
This adds an HTTP/2 304 "Not Modified" test to verify that we handle it
correctly.

This fixes #3147

mlibbey commented:
- Should we add a range request that updates a stale object?
- with h2 expected to have more requests per stream, should we test multiple requests?
- Should we add any cases from https://cache-tests.fyi/#update304 ?

maskit commented:
Having more tests is a good thing so I'm not going to block this, however, I wouldn't put the test under h2 directory because 304 is not an H2 specific thing. I wrote this on some previous PR before, but I think we should put tests under h2 directory only if those check H2 specific behavior. Other tests should be under other directories accordingly and the tests should be ran on all the HTTP versions.

bneradt commented:
> Having more tests is a good thing so I'm not going to block this, however, I wouldn't put the test under h2 directory because 304 is not an H2 specific thing. I wrote this on some previous PR before, but I think we should put tests under h2 directory only if those check H2 specific behavior. Other tests should be under other directories accordingly and the tests should be ran on all the HTTP versions.

Sounds good. I've moved the test to under cachedIMSRange.test.py which has other 304 tests with it already. I've also added HTTP and HTTPS versions of this test as well.

Thanks for the review.

bneradt commented:
> * Should we add a range request that updates a stale object?
> * with h2 expected to have more requests per stream, should we test multiple requests?
> * Should we add any cases from https://cache-tests.fyi/#update304 ?

These are good ideas. Thank you. For this PR, however, I'd like to stick to just this more narrow coverage. This PR is made to address a particular test need described in #3147 which couldn't be satisfied at the time due to a deficiency in curl's HTTP/2 read behavior.

maskit reviewed (approved):
Looks good to me, but I'm not sure if we can close #3147. The motivation of #3147 seems like to be testing 304 response with an empty body.

maskit commented:
Use of Proxy Verifier was suggested. Is it possible to test the case with curl now?

bneradt commented:
> Looks good to me, but I'm not sure if we can close #3147. The motivation of #3147 seems like to be testing 304 response with an empty body.

The 304 from ATS has an empty body.

> Use of Proxy Verifier was suggested. Is it possible to test the case with curl now?

Yes, Proxy Verifier was suggested because curl wasn't dealing with the end of the stream appropriately. But it seems to be now. When I add debug to the curl I'm using, I see that the on_stream_close callback is appropriately called.
